clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -4620,6 +4620,205 @@ bool SemaHLSL::transformInitList(const InitializedEntity &Entity,
+            << StringRef(&Chunk[0], 1)
+            << StringRef(Expected, sizeof(Expected) - 1)
+            << SourceRange(CompLoc);
+        return QualType();
+      }
+      RowChar = Chunk[1];
+      ColChar = Chunk[2];
+    }
+
+    // Must be digits.
+    bool isDigitsError = false;
+    if (!isDigit(RowChar)) {
+      const char Expected[] = "row as integer";
+      unsigned BadPos = IsZeroBasedAccessor ? 2 : 1;
+      S.Diag(OpLoc.getLocWithOffset(I + BadPos + 1),
+             diag::err_builtin_matrix_invalid_member)
+          << StringRef(&RowChar, 1) << StringRef(Expected, sizeof(Expected) - 1)
+          << SourceRange(CompLoc);
+      isDigitsError = true;
+    }
+
+    if (!isDigit(ColChar)) {
+      const char Expected[] = "column as integer";
+      unsigned BadPos = IsZeroBasedAccessor ? 3 : 2;
+      S.Diag(OpLoc.getLocWithOffset(I + BadPos + 1),
+             diag::err_builtin_matrix_invalid_member)
+          << StringRef(&ColChar, 1) << StringRef(Expected, sizeof(Expected) - 1)
+          << SourceRange(CompLoc);
+      isDigitsError = true;
+    }
+    if (isDigitsError)
+      return QualType();
+
+    unsigned Row = RowChar - '0';
+    unsigned Col = ColChar - '0';
+
+    bool HasIndexingError = false;
+    if (IsZeroBasedAccessor) {
+      // 0-based [0..3]
+      if (!isZeroBasedIndex(Row)) {
+        S.Diag(OpLoc, diag::err_hlsl_matrix_element_not_in_bounds)
+            << /*row*/ 0 << /*zero-based*/ 0 << SourceRange(CompLoc);
+        HasIndexingError = true;
+      }
+      if (!isZeroBasedIndex(Col)) {
+        S.Diag(OpLoc, diag::err_hlsl_matrix_element_not_in_bounds)
+            << /*col*/ 1 << /*zero-based*/ 0 << SourceRange(CompLoc);
+        HasIndexingError = true;
+      }
+    } else {
+      // 1-based [1..4]
+      if (!isOneBasedIndex(Row)) {
+        S.Diag(OpLoc, diag::err_hlsl_matrix_element_not_in_bounds)
+            << /*row*/ 0 << /*one-based*/ 1 << SourceRange(CompLoc);
+        HasIndexingError = true;
+      }
+      if (!isOneBasedIndex(Col)) {
+        S.Diag(OpLoc, diag::err_hlsl_matrix_element_not_in_bounds)
+            << /*col*/ 1 << /*one-based*/ 1 << SourceRange(CompLoc);
+        HasIndexingError = true;
+      }
+      // Convert to 0-based after range checking.
+      Row--;
+      Col--;
+    }
+
+    if (HasIndexingError)
+      return QualType();
+
+    // Note: matrix swizzle index is hard coded. That means Row and Col can
+    // potentially be larger than Rows and Cols if matrix size is less than
+    // the max index size.
+    bool HasBoundsError = false;
+    if (Row >= Rows) {
+      Diag(OpLoc, diag::err_hlsl_matrix_index_out_of_bounds)
+          << /*Row*/ 0 << Row << Rows << SourceRange(CompLoc);
+      HasBoundsError = true;
+    }
+    if (Col >= Cols) {
+      Diag(OpLoc, diag::err_hlsl_matrix_index_out_of_bounds)
+          << /*Col*/ 1 << Col << Cols << SourceRange(CompLoc);
+      HasBoundsError = true;
+    }
```

Comment:
```suggestion
    HasError |= validateMatrixIndex(Col, Cols, OpLoc, diag::err_hlsl_matrix_index_out_of_bounds, CompLoc, true /*IsColumn*/);
```
I'm wondering if it would make sense to consolidate each of these index checks into a call that would look something like this.